### ISSUE TITLE
Wavelength unit conversion

### ISSUE THREAD

**Issue by brechmos-stsci** [astropy/specutils#196]
Need the ability to convert wavelength units. This should leverage astropy units etc.

(will be used by specviz)

**Comment by nmearl** [comment]
@robel

**Comment by robelgeda** [comment]
Nothing new to add here. This feature is already implemented. This ticket can be closed.

**Comment by brechmos-stsci** [comment]
I might have missed them, but are there tests for the wavelength unit conversions?  If not, that would be a great thing to add.

**Comment by nmearl** [comment]
[Yes, there are](https://github.com/astropy/specutils/blob/master/specutils/tests/test_spectrum1d.py#L54)

**Comment by robelgeda** [comment]
I have made a mistake here @brechmos-stsci and @nmearl. It would be nice to add `to_spectral_axes` just like there is a `to_flux` function for flux unit conversions. I was under the impression that the  [`with_spectral_unit`](https://github.com/astropy/specutils/blob/master/specutils/spectra/spectrum_mixin.py#L171) function satisfied this ticket. This is not correct because `with_spectral_unit` returns a new `Spectrum1D` object instead of converting the units of the spectral unit in question.

**Comment by eteq** [comment]
@robelgeda - can you clarify the motivation behind a `to_spectral_axis`?  Can't the user just do `spectrum.spectral_axis.to(u.<whatever>)`?

(Same question for `to_flux`, now that I think about it...)

**Comment by robelgeda** [comment]
I am not sure actually, @nmearl what are your thoughts?

**Comment by nmearl** [comment]
@eteq `spectrum.spectral_axis.to` requires that the units the spectrum has on its instance are correct and convertible. @robelgeda's proposal is to have methods that allow a user to _overwrite_ the current unit, without consideration of conversion. E.g. in the case of incorrect units in a fits header being read into a new `Spectrum1D` object.

**Comment by eteq** [comment]
Aha, now that I understand I am against this.  It's a lot more work to try to keep things validated when the user might change the unit at any time without any checks.  It's also heavily overloading the word `to` here, which in all other contexts means "give me a new object with where I *convert* the unit".  Users will *certainly* be confused by mixing those.

Instead to address this use case I'd say you'd do something like this:
```
new_spec = Spectrum1D(flux=old_spec.flux.value*newunit, spectral_axis=old_spec.spectral_axis)
```
That's more explicit anyway.

**Comment by hcferguson** [comment]
@eteq I agree that seems much safer. My only concern is whether this supports the SpecViz use case,
where one wants to change the units in the *view*. If changing the units always means returning a new instance, does that mean SpecViz will have to (or should, for the sake of consistency) create a new layer?

If you have only one layer this isn't such a big deal either way. But if you have a bunch of layers and want to change the units view on all of them, this could get inconvenient. 

I suppose SpecViz could just internally overwrite all the old instances with the new ones in a loop while changing the units, so probably it's not a big deal except for possible confusion when they write out a spectrum and find it's in the units of the view and not the original ones.

**Comment by hcferguson** [comment]
I'm of two minds...

When you change to a log scale or semi-log scale in matplotlib, for example, you aren't creating a new array (although, quite honestly, I usually do). 

If I'm in MOSViz, using SpecViz and SpecUtils under the hood to do the 1D visualization and computations, then what if I want to change the display of my flux units from nu to nu*fnu for all the spectra? Am I going to 
 1. get 100 new layers, which would be consistent with having unit changes always create a new instance,
 2. have the MOSViz overwrite the current spectra with versions that have the new units,
 3. or have MOSViz keep the spectra in the original units but *display them* in the currently selected units? 

At some level, I think we need to enable all three of these. For MOSViz, which is intended for quick look, (3) is going to be the most convenient.

That said, I agree with the suggestion to create a new instance in SpecUtils rather than change units in place. We will just have to be careful in MOSViz and SpecViz to deal with the complications of supporting both *changing the view* and *changing the data* and making it clear what's happening when. The implication of changing just the view might be that the function that does the plotting might have to accept as arguments functions (and any additional arguments) to operate on both axes to change the units. Although this sounds a bit messy and might affect performance. If we don't want support (3) then I think we at least need the option to support (2) in the viz tools, but that doesn't really have implications for what we do in SpecUtils.

**Comment by eteq** [comment]
> When you change to a log scale or semi-log scale in matplotlib, for example, you aren't creating a new array (although, quite honestly, I usually do).

Actually, I think this *does* create a new one under the hood.  Sort of more like the case 3 you mention above, @hcferguson. But I agree that leads to some of your concerns, so I'm inclined more towards your case 2 - then in a more-interactive version of MOSViz/specviz (i.e., where you can pop back into a more code-based workflow after doing some analysis with the gui tool), you have to specifically ask for the "modified" version of the spectrum if you want to have any units updated.

**Comment by eteq** [comment]
Switching this to 0.5 as it should be decided in parallel with #312

**Comment by nmearl** [comment]
Seems to me that this issue is resolved in #309 and follow up was discussed in #312. Feel free to reopen if this is not the case.
